### PR TITLE
feat(theming): add componentStyles utility

### PR DESCRIPTION
**Pull request by jzempel**
## Description

This is an improved version of `retrieveComponentStyles`, with:
- improved typing (consumer doesn't need to fuss with `data-` attribute exceptions in styled components)
- the ability to specify a custom ID attribute (defaults to `data-garden-id`)
- defaults support no-parameter implementations
- naming aligns better with other Garden view component conventions (`sizeStyles` & `colorStyles`)

## Detail

All instances of the `@deprecated retrieveComponentStyles` will be replaced in a follow-on PR.

## Checklist

- [ ] :ok_hand: ~design updates will be Garden Designer approved (add the designer as a reviewer)~
- [x] :globe_with_meridians: demo is up-to-date (`npm start`)
- [ ] :arrow_left: ~renders as expected with reversed (RTL) direction~
- [ ] :black_circle: ~renders as expected in dark mode~
- [ ] :metal: ~renders as expected with Bedrock CSS (`?bedrock`)~
- [x] :guardsman: includes new unit tests. Maintain existing coverage (always >= 96%)
- [ ] :wheelchair: ~tested for WCAG 2.1 AA accessibility compliance~
- [ ] :memo: ~tested in Chrome, Firefox, Safari, and Edge~

### CODE REVIEW

File: packages/theming/src/utils/componentStyles.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * CSS for component customizations based on `theme.components[componentId]`.
+ *
+ * @param {Object} props.theme Provides `components` object use to resolve the given component ID
+ * @param {String} [props.componentId] Specifies the lookup id for * `theme.components` styles.
+ *  The ID will be inferred from the `'data-garden-id'` attribute if not provided.
+ *
+ * @returns component CSS styles
+ */
+export const componentStyles = (props: ThemeProps<DefaultTheme> & { componentId?: string }) => {
```
Comment (ze-flo) [inline packages/theming/src/utils/componentStyles.ts:19]: ThemeProps was dropped in `styled-component` v6. To reduce the refactoring scope within our internal `styled` components, I [added back](https://github.com/zendeskgarden/react-components/blame/main/utils/build/styled.d.ts#L16-L18) `ThemeProps` to the `styled-components` namespace. However, `ThemeProps` won't be available for users on `styled-components` v6.

Here's the built output for `componentStyles.d.ts` showing the missing type with v6. 

Screenshot 2024-12-06 at 11 00 19 AM

It's probably best to explicitly define the `theme` prop.

```suggestion
export const componentStyles = (props: { componentId?: string, theme: DefaultTheme }) => {
```

File: packages/theming/src/utils/retrieveComponentStyles.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 import { ThemeProps, DefaultTheme } from 'styled-components';
 
-/** @component */
+/** @deprecated Use `componentStyles` instead. */
```
Comment (ze-flo) [inline packages/theming/src/utils/retrieveComponentStyles.ts:10]: Is `/** @component */` needed for the documentation?

File: packages/theming/src/utils/retrieveComponentStyles.ts [hunk 3]
```diff
@@ -7,7 +7,7 @@
...
 export default function retrieveComponentStyles(
   componentId: string,
   props: Partial<ThemeProps<Partial<DefaultTheme>>>
```
Comment (ze-flo) [inline packages/theming/src/utils/retrieveComponentStyles.ts:13]: This is an oversight from #1978. 😞 
```suggestion
  props: { theme?: <Partial<DefaultTheme> }
```

File: packages/theming/src/utils/componentStyles.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  let retVal: string | undefined;
+  const components = props.theme?.components;
```
Comment (ze-flo) [inline packages/theming/src/utils/componentStyles.ts:21]: `ThemeProps` (or my suggestion ☝️) defines `theme` as [required](https://github.com/zendeskgarden/react-components/blob/main/utils/build/styled.d.ts). If this is what we want going forward:  
```suggestion
  const components = props.theme.components;
```

**Comment by ze-flo** [comment]
Once last thing I forgot to mention. 😞 Can we update the [template](https://github.com/zendeskgarden/react-components/blob/main/packages/.template/src/styled/Styled%7B%7Bcapitalize%20component%7D%7D.ts#L81) for StyledComponents?

**Comment by jzempel** [comment]
> Once last thing I forgot to mention. 😞 Can we update the [template](https://github.com/zendeskgarden/react-components/blob/main/packages/.template/src/styled/Styled%7B%7Bcapitalize%20component%7D%7D.ts#L81) for StyledComponents?

I'll get that in the follow-on PR mentioned in the description :wink:

**Review by ze-flo — APPROVED** [review]
Nice! 💯

**Pull request merged**